File: raspberry-pi/c/cat9555.c
/***********************************
 * Seengreat CAT9555 IO Expansion Module demo
 * Author(s):Andy Li from Seengreat
 * ********************************/

#include <string.h>
#include <stdio.h>
#include <time.h>
#include <wiringPi.h>
#include <wiringPiI2C.h>

#define  DEV_ADRR   0x27  //when A2=1 A1=1 A0=1 =>0010 0111

//CAT9555 regsiter
#define IN_PORT0    0
#define IN_PORT1    1
#define OUT_PORT0   2
#define OUT_PORT1   3
#define PI_PORT0    4
#define PI_PORT1    5
#define CFG_PORT0   6
#define CFG_PORT1   7

#define PIN_INT 7   //GPIO.7 in wiringPi encode

int i2c_fd = 0;
char p0_val = 0x00;
char p1_val = 0xff;
unsigned char int_flag = 0;
unsigned char res = 0;
unsigned int t = 0;
/****************************************************
 * name：WriteRawRC   
 * describe：write RC522 register  
 * input ：ucAddress， ucValue
 */
void Write_reg( unsigned char reg_addr, unsigned char value )
{  
    wiringPiI2CWriteReg8(i2c_fd, reg_addr&0x07, value );
}
/***************************************************************
 * name：ReadRawRC 
 * read current value from the register
 */
unsigned char Read_reg ( unsigned char reg_addr )
{
    return wiringPiI2CReadReg8(i2c_fd, reg_addr&0x07);
}

void Read_pin_state(void)
{
    if(int_flag == 0)
    {
       int_flag = 1;    
    }
}

void Cat9555_Init(void)
{
    wiringPiSetup(); //init wiringpi lib
    i2c_fd = wiringPiI2CSetup(DEV_ADRR);
    printf("dev_addr:%d\r\n", i2c_fd);
    pinMode(PIN_INT,INPUT);
    pullUpDnControl(PIN_INT,PUD_UP);
    wiringPiISR(PIN_INT,INT_EDGE_FALLING, &Read_pin_state);
}


int main(void)
{
    Cat9555_Init();
    Write_reg(CFG_PORT0, 0x00);//set the port 0 as output
    Write_reg(CFG_PORT1, 0xFF);//set the port 1 as intput
    printf("cfg_port0:%x \r\n", Read_reg(CFG_PORT0));//read the configuration port 0 register
    printf("cfg_port1:%x \r\n", Read_reg(CFG_PORT1));//read the configuration port 1 register
    Write_reg(PI_PORT1, 0x00);
    while(1)
    {
        delay(20);
        t++;
        if(t>49)//Invert the level of the P0 for about 1 second(20*50=1000ms)
        {
            t = 0;
            p0_val = ~p0_val;
            Write_reg(OUT_PORT0, p0_val);            
        } 
        if(int_flag == 1) //check the interrupt pin state
        {
            res = Read_reg(IN_PORT1);
            if(p1_val != res)//The level of P1 has changed
            {
                p1_val = res;
                printf("port 1:%x\r\n",p1_val);
                int_flag = 0;
            }
        }
    }
    
}

File: raspberry-pi/python/cat9555.py
# Seengreat CAT9555 IO Expansion Module demo
# Author(s):Andy Li from Seengreat

import smbus
import time
import numpy as np
import wiringpi as wpi

#CAT9555 regsiter
IN_PORT0     = 0
IN_PORT1     = 1
OUT_PORT0    = 2
OUT_PORT1    = 3
PI_PORT0     = 4
PI_PORT1     = 5
CFG_PORT0    = 6
CFG_PORT1    = 7

#CAT9555 address
#0 1 0 0 A2 A1 A0 R/W
#R/W bit = 0 is write command
#R/W bit = 1 is read  command

PIN_INT = 7  # GPIO.7 in wiringPi encode

class CAT9555_api():
    def __init__(self):
        self.dev_addr = 0x20 #when A2=0 A1=0 A0=0 =>0010 0000
        self.i2c = smbus.SMBus(1)  # /dev/i2c-1
        self.port1 = 0xff
        wpi.wiringPiSetup()
        wpi.pinMode(PIN_INT,wpi.INPUT)
        wpi.pullUpDnControl(PIN_INT,wpi.PUD_UP)
        wpi.wiringPiISR(PIN_INT,wpi.INT_EDGE_FALLING, self.read_pin_state)
        time.sleep(1)
    def write_reg (self, reg_addr, value):
        """"write cat9555 register"""
        self.i2c.write_byte_data(self.dev_addr, reg_addr&0x07, value)
    def read_reg (self, reg_addr):
        """read current value from the register"""
        res = self.i2c.read_byte_data(self.dev_addr, reg_addr&0x07)
        return res
    def read_pin_state(self):
        time.sleep(0.02)
        res = cat9555.read_reg(IN_PORT1)
        if self.port1 != res:
            self.port1 = res
            print("port 1:",hex(self.port1))
        return 1
    
if __name__ == '__main__':
    cat9555 = CAT9555_api()
    cat9555.write_reg(CFG_PORT0, 0X00)  # set the port 0 as output
    cat9555.write_reg(CFG_PORT1, 0XFF)  # set the port 1 as intput
    res = cat9555.read_reg(CFG_PORT0)  # read the configuration port 0 register
    print("cfg_port0:", hex(res))
    res = cat9555.read_reg(CFG_PORT1)  # read the configuration port 1 register
    print("cfg_port1:", hex(res))
    cat9555.write_reg(PI_PORT1, 0X00)
    p0_val = 0x00
    while True:
        p0_val = ~p0_val
        cat9555.write_reg(OUT_PORT0, p0_val)
        time.sleep(1)
              
        
        
        
